src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -178,15 +178,123 @@ void Compiler::optRelopImpliesRelop(RelopImplicationInfo* rii)
 
     genTreeOps const oper = genTreeOps(funcApp.m_func);
 
+    // Exclude floating point relops.
+    //
+    if (varTypeIsFloating(vnStore->TypeOfVN(funcApp.m_args[0])))
+    {
+        return;
+    }
+
+#ifdef DEBUG
+    static ConfigMethodRange JitEnableRboRange;
+    JitEnableRboRange.EnsureInit(JitConfig.JitEnableRboRange());
+    const unsigned hash    = impInlineRoot()->info.compMethodHash();
+    const bool     inRange = JitEnableRboRange.Contains(hash);
+#else
+    const bool inRange = true;
+#endif
+
+    // Dominating compare has the form R(x,y)
+    // See if tree compare has the form R*(x,y) or R*(y,x) where we can infer R* from R
+    //
+    // Could also extend to the unsigned VN relops.
+    //
+    VNFuncApp treeApp;
+    if (inRange && GenTree::OperIsCompare(oper) && vnStore->GetVNFunc(rii->treeNormVN, &treeApp))
+    {
+        genTreeOps const treeOper = genTreeOps(treeApp.m_func);
+        genTreeOps       domOper  = oper;
+
+        if (((treeApp.m_args[0] == funcApp.m_args[0]) && (treeApp.m_args[1] == funcApp.m_args[1])) ||
+            ((treeApp.m_args[0] == funcApp.m_args[1]) && (treeApp.m_args[1] == funcApp.m_args[0])))
+        {
+            const bool swapped = (treeApp.m_args[0] == funcApp.m_args[1]);
+
+            if (swapped)
+            {
+                domOper = GenTree::SwapRelop(domOper);
+            }
+
+            switch (domOper)
+            {
+                default:
+                    break;
+
+                case GT_EQ:
+                    rii->canInfer =
+                        (treeOper == GT_GE) || (treeOper == GT_LE) || (treeOper == GT_GT) || (treeOper == GT_LT);
+                    rii->reverseSense      = ((treeOper == GT_GT) || (treeOper == GT_LT));
+                    rii->canInferFromTrue  = true;
+                    rii->canInferFromFalse = false;
+                    rii->vnRelation        = ValueNumStore::VN_RELATION_KIND::VRK_Unrelated;
```

Comment:
I am trying to understand the meaning of `VRK_Unrelated` here. What we are saying is although we can infer R* from R, they are not related meaning they cannot be simply swapped or reversed, but they have slightly different operators. Can we have something meaningful enum value for such cases like `VRK_DomInferred` or similar?

Reply:
Happy to rename it if you think it makes things clearer.